File: OpenHome/Net/ControlPoint/Proxies/CpAvOpenhomeOrgUSBInfo1.cpp
#include "CpAvOpenhomeOrgUSBInfo1.h"
#include <OpenHome/Net/Core/CpProxy.h>
#include <OpenHome/Net/Private/CpiService.h>
#include <OpenHome/Private/Thread.h>
#include <OpenHome/Net/Private/AsyncPrivate.h>
#include <OpenHome/Net/Core/CpDevice.h>
#include <OpenHome/Net/Private/Error.h>
#include <OpenHome/Net/Private/CpiDevice.h>

using namespace OpenHome;
using namespace OpenHome::Net;


class SyncGetUSBInfoAvOpenhomeOrgUSBInfo1 : public SyncProxyAction
{
public:
    SyncGetUSBInfoAvOpenhomeOrgUSBInfo1(CpProxyAvOpenhomeOrgUSBInfo1& aProxy, TBool& aUSBPlugStatus, Brh& aUSBERRORInfo, Brh& aUSBCapacity, Brh& aUSBFileCount);
    virtual void CompleteRequest(IAsync& aAsync);
    virtual ~SyncGetUSBInfoAvOpenhomeOrgUSBInfo1() {}
private:
    CpProxyAvOpenhomeOrgUSBInfo1& iService;
    TBool& iUSBPlugStatus;
    Brh& iUSBERRORInfo;
    Brh& iUSBCapacity;
    Brh& iUSBFileCount;
};

SyncGetUSBInfoAvOpenhomeOrgUSBInfo1::SyncGetUSBInfoAvOpenhomeOrgUSBInfo1(CpProxyAvOpenhomeOrgUSBInfo1& aProxy, TBool& aUSBPlugStatus, Brh& aUSBERRORInfo, Brh& aUSBCapacity, Brh& aUSBFileCount)
    : iService(aProxy)
    , iUSBPlugStatus(aUSBPlugStatus)
    , iUSBERRORInfo(aUSBERRORInfo)
    , iUSBCapacity(aUSBCapacity)
    , iUSBFileCount(aUSBFileCount)
{
}

void SyncGetUSBInfoAvOpenhomeOrgUSBInfo1::CompleteRequest(IAsync& aAsync)
{
    iService.EndGetUSBInfo(aAsync, iUSBPlugStatus, iUSBERRORInfo, iUSBCapacity, iUSBFileCount);
}


CpProxyAvOpenhomeOrgUSBInfo1::CpProxyAvOpenhomeOrgUSBInfo1(CpDevice& aDevice)
    : CpProxy("av-openhome-org", "USBInfo", 1, aDevice.Device())
{
    OpenHome::Net::Parameter* param;

    iActionGetUSBInfo = new Action("GetUSBInfo");
    param = new OpenHome::Net::ParameterBool("USBPlugStatus");
    iActionGetUSBInfo->AddOutputParameter(param);
    param = new OpenHome::Net::ParameterString("USBERRORInfo");
    iActionGetUSBInfo->AddOutputParameter(param);
    param = new OpenHome::Net::ParameterString("USBCapacity");
    iActionGetUSBInfo->AddOutputParameter(param);
    param = new OpenHome::Net::ParameterString("USBFileCount");
    iActionGetUSBInfo->AddOutputParameter(param);
}

CpProxyAvOpenhomeOrgUSBInfo1::~CpProxyAvOpenhomeOrgUSBInfo1()
{
    DestroyService();
    delete iActionGetUSBInfo;
}

void CpProxyAvOpenhomeOrgUSBInfo1::SyncGetUSBInfo(TBool& aUSBPlugStatus, Brh& aUSBERRORInfo, Brh& aUSBCapacity, Brh& aUSBFileCount)
{
    SyncGetUSBInfoAvOpenhomeOrgUSBInfo1 sync(*this, aUSBPlugStatus, aUSBERRORInfo, aUSBCapacity, aUSBFileCount);
    BeginGetUSBInfo(sync.Functor());
    sync.Wait();
}

void CpProxyAvOpenhomeOrgUSBInfo1::BeginGetUSBInfo(FunctorAsync& aFunctor)
{
    Invocation* invocation = iService->Invocation(*iActionGetUSBInfo, aFunctor);
    TUint outIndex = 0;
    const Action::VectorParameters& outParams = iActionGetUSBInfo->OutputParameters();
    invocation->AddOutput(new ArgumentBool(*outParams[outIndex++]));
    invocation->AddOutput(new ArgumentString(*outParams[outIndex++]));
    invocation->AddOutput(new ArgumentString(*outParams[outIndex++]));
    invocation->AddOutput(new ArgumentString(*outParams[outIndex++]));
    iInvocable.InvokeAction(*invocation);
}

void CpProxyAvOpenhomeOrgUSBInfo1::EndGetUSBInfo(IAsync& aAsync, TBool& aUSBPlugStatus, Brh& aUSBERRORInfo, Brh& aUSBCapacity, Brh& aUSBFileCount)
{
    ASSERT(((Async&)aAsync).Type() == Async::eInvocation);
    Invocation& invocation = (Invocation&)aAsync;
    ASSERT(invocation.Action().Name() == Brn("GetUSBInfo"));

    Error::ELevel level;
    TUint code;
    const TChar* ignore;
    if (invocation.Error(level, code, ignore)) {
        THROW_PROXYERROR(level, code);
    }
    TUint index = 0;
    aUSBPlugStatus = ((ArgumentBool*)invocation.OutputArguments()[index++])->Value();
    ((ArgumentString*)invocation.OutputArguments()[index++])->TransferTo(aUSBERRORInfo);
    ((ArgumentString*)invocation.OutputArguments()[index++])->TransferTo(aUSBCapacity);
    ((ArgumentString*)invocation.OutputArguments()[index++])->TransferTo(aUSBFileCount);
}

File: OpenHome/Net/Bindings/Cs/ControlPoint/Proxies/CpAvOpenhomeOrgUSBInfo1.cs
using System;
using System.Runtime.InteropServices;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using OpenHome.Net.Core;
using OpenHome.Net.ControlPoint;

namespace OpenHome.Net.ControlPoint.Proxies
{
    public interface ICpProxyAvOpenhomeOrgUSBInfo1 : ICpProxy, IDisposable
    {
        void SyncGetUSBInfo(out bool aUSBPlugStatus, out String aUSBERRORInfo, out String aUSBCapacity, out String aUSBFileCount);
        void BeginGetUSBInfo(CpProxy.CallbackAsyncComplete aCallback);
        void EndGetUSBInfo(IntPtr aAsyncHandle, out bool aUSBPlugStatus, out String aUSBERRORInfo, out String aUSBCapacity, out String aUSBFileCount);
    }

    internal class SyncGetUSBInfoAvOpenhomeOrgUSBInfo1 : SyncProxyAction
    {
        private CpProxyAvOpenhomeOrgUSBInfo1 iService;
        private bool iUSBPlugStatus;
        private String iUSBERRORInfo;
        private String iUSBCapacity;
        private String iUSBFileCount;

        public SyncGetUSBInfoAvOpenhomeOrgUSBInfo1(CpProxyAvOpenhomeOrgUSBInfo1 aProxy)
        {
            iService = aProxy;
        }
        public bool USBPlugStatus()
        {
            return iUSBPlugStatus;
        }
        public String USBERRORInfo()
        {
            return iUSBERRORInfo;
        }
        public String USBCapacity()
        {
            return iUSBCapacity;
        }
        public String USBFileCount()
        {
            return iUSBFileCount;
        }
        protected override void CompleteRequest(IntPtr aAsyncHandle)
        {
            iService.EndGetUSBInfo(aAsyncHandle, out iUSBPlugStatus, out iUSBERRORInfo, out iUSBCapacity, out iUSBFileCount);
        }
    };

    /// <summary>
    /// Proxy for the av.openhome.org:USBInfo:1 UPnP service
    /// </summary>
    public class CpProxyAvOpenhomeOrgUSBInfo1 : CpProxy, IDisposable, ICpProxyAvOpenhomeOrgUSBInfo1
    {
        private OpenHome.Net.Core.Action iActionGetUSBInfo;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <remarks>Use CpProxy::[Un]Subscribe() to enable/disable querying of state variable and reporting of their changes.</remarks>
        /// <param name="aDevice">The device to use</param>
        public CpProxyAvOpenhomeOrgUSBInfo1(CpDevice aDevice)
            : base("av-openhome-org", "USBInfo", 1, aDevice)
        {
            OpenHome.Net.Core.Parameter param;
            List<String> allowedValues = new List<String>();

            iActionGetUSBInfo = new OpenHome.Net.Core.Action("GetUSBInfo");
            param = new ParameterBool("USBPlugStatus");
            iActionGetUSBInfo.AddOutputParameter(param);
            param = new ParameterString("USBERRORInfo", allowedValues);
            iActionGetUSBInfo.AddOutputParameter(param);
            param = new ParameterString("USBCapacity", allowedValues);
            iActionGetUSBInfo.AddOutputParameter(param);
            param = new ParameterString("USBFileCount", allowedValues);
            iActionGetUSBInfo.AddOutputParameter(param);
        }

        /// <summary>
        /// Invoke the action synchronously
        /// </summary>
        /// <remarks>Blocks until the action has been processed
        /// on the device and sets any output arguments</remarks>
        /// <param name="aUSBPlugStatus"></param>
        /// <param name="aUSBERRORInfo"></param>
        /// <param name="aUSBCapacity"></param>
        /// <param name="aUSBFileCount"></param>
        public void SyncGetUSBInfo(out bool aUSBPlugStatus, out String aUSBERRORInfo, out String aUSBCapacity, out String aUSBFileCount)
        {
            SyncGetUSBInfoAvOpenhomeOrgUSBInfo1 sync = new SyncGetUSBInfoAvOpenhomeOrgUSBInfo1(this);
            BeginGetUSBInfo(sync.AsyncComplete());
            sync.Wait();
            sync.ReportError();
            aUSBPlugStatus = sync.USBPlugStatus();
            aUSBERRORInfo = sync.USBERRORInfo();
            aUSBCapacity = sync.USBCapacity();
            aUSBFileCount = sync.USBFileCount();
        }

        /// <summary>
        /// Invoke the action asynchronously
        /// </summary>
        /// <remarks>Returns immediately and will run the client-specified callback when the action
        /// later completes.  Any output arguments can then be retrieved by calling
        /// EndGetUSBInfo().</remarks>
        /// <param name="aCallback">Delegate to run when the action completes.
        /// This is guaranteed to be run but may indicate an error</param>
        public void BeginGetUSBInfo(CallbackAsyncComplete aCallback)
        {
            Invocation invocation = iService.Invocation(iActionGetUSBInfo, aCallback);
            int outIndex = 0;
            invocation.AddOutput(new ArgumentBool((ParameterBool)iActionGetUSBInfo.OutputParameter(outIndex++)));
            invocation.AddOutput(new ArgumentString((ParameterString)iActionGetUSBInfo.OutputParameter(outIndex++)));
            invocation.AddOutput(new ArgumentString((ParameterString)iActionGetUSBInfo.OutputParameter(outIndex++)));
            invocation.AddOutput(new ArgumentString((ParameterString)iActionGetUSBInfo.OutputParameter(outIndex++)));
            iService.InvokeAction(invocation);
        }

        /// <summary>
        /// Retrieve the output arguments from an asynchronously invoked action.
        /// </summary>
        /// <remarks>This may only be called from the callback set in the above Begin function.</remarks>
        /// <param name="aAsyncHandle">Argument passed to the delegate set in the above Begin function</param>
        /// <param name="aUSBPlugStatus"></param>
        /// <param name="aUSBERRORInfo"></param>
        /// <param name="aUSBCapacity"></param>
        /// <param name="aUSBFileCount"></param>
        public void EndGetUSBInfo(IntPtr aAsyncHandle, out bool aUSBPlugStatus, out String aUSBERRORInfo, out String aUSBCapacity, out String aUSBFileCount)
        {
            uint code;
            string desc;
            if (Invocation.Error(aAsyncHandle, out code, out desc))
            {
                throw new ProxyError(code, desc);
            }
            uint index = 0;
            aUSBPlugStatus = Invocation.OutputBool(aAsyncHandle, index++);
            aUSBERRORInfo = Invocation.OutputString(aAsyncHandle, index++);
            aUSBCapacity = Invocation.OutputString(aAsyncHandle, index++);
            aUSBFileCount = Invocation.OutputString(aAsyncHandle, index++);
        }

        /// <summary>
        /// Must be called for each class instance.  Must be called before Core.Library.Close().
        /// </summary>
        public void Dispose()
        {
            lock (this)
            {
                if (iHandle == IntPtr.Zero)
                    return;
                DisposeProxy();
                iHandle = IntPtr.Zero;
            }
            iActionGetUSBInfo.Dispose();
        }
    }
}

